### DISCUSSION TITLE
Use a SolidJS store to make selectors rerun only when needed, automatically

### DISCUSSION THREAD

**Discussion by guillaumebrunerie** [reduxjs/react-redux#2043] [Ideas]
I've been thinking about Redux performance lately, and in the type of apps I am building (large and complex UI, large and complex state, and a lot of places where dragging something with the mouse dispatches an action at every `mousemove`), selectors seem to affect performance significantly, even though I spent a lot of time and thought on memoization. It is also difficult to get memoization right, as the various discussions in the Reselect repository attest.

But more precisely, I feel like the fact that **_every single selector reruns at every single state change_** is a performance bottleneck that is hard to overcome when an app gets big enough (also, it makes memoization critical). If you make small components each selecting their own data (which I believe is the recommended way to do, as it reduces the cost of React rerenders), you can relatively easily get thousands or tens of thousands of selectors that need to rerun at every frame, so even if they are all very fast/memoized, it will still take a non-negligible amount of time.

After reading more about SolidJS, I’m starting to really like the idea of signals and of reactive state, and it occurred to me that we could actually reuse Solid's stores and effects (but keep React for rendering) to get a pretty different implementation of React-Redux, with essentially the same API it has now, but with selectors that would rerun only when needed, and without having to memoize anything explicitly.

The idea is:
- make the global state into in a SolidJS store (which is an object wrapped in a proxy to make reactivity work)
- for `dispatch`, use the store's `setState` together with Solid's `produce` function, which means that you simply need to give a function that "mutates" the draft passed as an argument, exactly like we do today with Immer/RTK
- for `useSelector`, create a Solid effect that applies the selector to the global state and store the result in a regular React `useState`. An effect in Solid is a reactive function that remember exactly which parts of the global state were accessed and will rerun automatically every time one of those parts gets modified by a dispatched action.

I wonder if someone already explored a similar approach? What do you think about it? Are there some drawbacks that I am not thinking about?

Note that while it is also using proxies, this is pretty different from the ideas of `proxy-memoize`. If I understand it correctly, `proxy-memoize` "only" intends to make selectors faster by memoizing them automatically, but the resulting memoized selectors still need to run after every dispatched action (as this is handled by `react-redux`). In contrast, what I am proposing would skip running irrelevant selectors entirely, so there would be absolutely no performance penalty in having many selectors that access unchanged parts of the state. And just like `proxy-memoize`, it would also simplify memoization considerably as you mostly wouldn’t need to think about it at all.

---

Here is a proof of concept, a drop-in replacement for basic versions of `createReducer`, `configureStore`, `Provider`, `useDispatch`, and `useSelector` (and you can keep using `createAction` from RTK). It is less than 100 lines of codes (including comments) and I tested it in a toy todo-list example. It works the way is described above, for instance toggling a todo item only reruns the selectors that actually look as the `completed` property of the given todo item, and the other selectors do not rerun at all.

```javascript
import * as React from "react";
import { createEffect, createRoot, untrack } from "solid-js";
import { createStore, produce } from "solid-js/store";

// I’m kind of abusing `createReducer` here in that it doesn’t actually create a
// reducer but simply collects the reducers and the initial state
export const createReducer = (initialState, builderCallback) => {
    const reducers = {};
    const builder = {
        addCase: (actionCreator, reducer) => {
            reducers[actionCreator.type] = reducer;
            return builder;
        }
    };
    builderCallback(builder);
    return {
        initialState,
        reducers,
    };
}

// `configureStore` does not take a reducer as an argument, but takes whatever
// is returned by `createReducer`
export const configureStore = ({reducer: {initialState, reducers}}) => {
    // `createStore` comes from Solid, it wraps the state into a proxy so that
    // it will be able to track which parts are accessed by selectors
    const [state, setState] = createStore(initialState);
    const getState = () => state;
    // `dispatch` simply uses Solid’s `produce` function, which is inspired by
    // Immer. The function contains mutating logic, with the proxy keeping track
    // of what changed.
    const dispatch = (action) => {
        setState(produce(state => {
            reducers[action.type](state, action);
        }));
    }
    const store = {getState, dispatch};
    return store;
};

const StoreContext = React.createContext(null);
export const Provider = ({store, children}) => {
    return (
        <StoreContext.Provider value={store}>
            {children}
        </StoreContext.Provider>
    )
};

export const useDispatch = () => {
    const store = React.useContext(StoreContext);
    return store.dispatch;
};

export const useSelector = (selector) => {
    const store = React.useContext(StoreContext);
    const state = store.getState();
    // When running for the first time, we call the selector on the state, but
    // it does not need to be tracked by Solid as we do it later in the effect
    const [selectedValue, setSelectedValue] = React.useState(
        () => untrack(() => selector(state))
    );
    React.useEffect(() => {
        // A root is used to make sure the Solid effect gets cleaned up properly
        return createRoot((dispose) => {
            let disposed = false;
            // `createEffect` comes from Solid, the given callback will
            // automatically detect which parts of `state` are used by
            // `selector(state)` and rerun if one of them changes. In particular
            // it will automatically update the React state when needed.
            createEffect(() => {
                try {
                    setSelectedValue(selector(state));
                } catch (e) {
                    // Similar trick as in React-Redux to deal with zombie
                    // children, if it throws an error it might be because of
                    // zombie children, so just try again a bit later
                    setTimeout(() => {
                        if (!disposed) {
                            setSelectedValue(selector(state));
                        };
                    });
                }
            });
            return () => {
                disposed = true;
                dispose();
            };
        });
    }, [state]);
    return selectedValue;
};
```

**Comment by markerikson** [comment]
This is certainly an idea I'd like to experiment with in general :) Actually just talked with some folks from Slack in the last week who are seeing 10-20K selectors running on each dispatch, and that itself is a sizeable bottleneck.

I do have some concerns with the example implementation, though.

My most immediate concern with using Solid here is that just importing and using those methods is 10.5K min / 4.5K min+gz. 

Second, I don't think we can realistically do anything that involves changing `configureStore`.  The POC work I've done with Reselect borrows from `tracked-redux` and does reconciliation of the resulting state.  But, from there I have concerns about dealing with large states.

**Comment by guillaumebrunerie** [comment]
Right, it's definitely only meant as a proof of concept for now, but I think it's very interesting to compare the approaches of Redux and Solid.

I think that the main difference can be illustrated pretty nicely using the concept of memoization. Let's say you have a pure function `fn: (t: T) => U` that you wish to memoize. Without any other assumption it's basically impossible, because the argument could have been mutated in arbitrary ways, so there is no way to memoize it efficiently.

Redux' approach: Assume the argument to be immutable. Now it's easy to memoize, because you know that if the argument didn’t change, the result didn’t change either. You can implement such a memoization function using WeakMaps for instance:
```typescript
const memoizeImmutable = <T extends object, U>(fn: (t: T) => U) => {
    const cache = new WeakMap<T, U>();
    return (t: T): U => {
        if (!cache.has(t)) {
            cache.set(t, fn(t));
        }
        return cache.get(t);
    }
}
```
It basically caches the result of `f` for each argument that we encounter. WeakMaps are just used as an illustration, a similar logic is also used in `useSelector` to say "if the result did not change since last time, do not rerender".

Solid's approach: Assume the argument to be a Solid store, which means that it is wrapped in a proxy in such a way that every time it is modified, its subscribers will be notified. This is very different from the case above, because the argument is now typically always the same reference, being mutated in place. Nevertheless, it is easy to memoize as well because you can simply subscribe to changes and then you will be notified of every change relevant to the part you subscribed to, so you can keep the memoized value up to date. Here is again how to implement it using WeakMaps:
```typescript
const memoizeSolidStore = <T extends object, U>(fn: (t: T) => U) => {
    const cache = new WeakMap<T, () => U>();
    return (t: T): U => {
        if (!cache.has(t)) {
            cache.set(t, createRoot(() => createMemo(() => fn(t))));
        }
        return cache.get(t)();
    }
}
```
This time we do not store results in the cache (as the same argument (w.r.t. referential equality) may give different results), but instead we store a *reactive value*, which is simply a function `() => U` that subscribes to changes to the parts of `t` used in `fn(t)`, and then always returns an up to date value of the result. The call to `createMemo` creates such a reactive value, it subscribes to the Solid store simply by calling `fn(t)` (accessing a value inside a store while in a reactive context like `createMemo` automatically subscribes to this value), and the call to `createRoot` is a technicality to make sure Solid doesn’t dispose of the subscription if it is called inside another reactive context.

Both approaches accomplish exactly the same thing: being able to mutate a value using imperative logic, but in a way that still makes it cheap to know whether some specific part of that value actually changed.
* In Redux, the imperative logic is converted to immutable logic using Immer, and then the check is simply a reference equality check.
* In Solid, the imperative logic stays as it is but the value is wrapped in a proxy that makes it so that the places that care about knowing whether a specific part of the value changed will be notified of changes.

But out of those two approaches, it seems like the Redux approach has more drawbacks, in particular it requires calling all selectors every time, and it is also potentially heavy on the GC as it creates a lot of new objects. The Solid approach, on the other hand, does not require calling irrelevant selectors (as relevant selectors will be notified directly from the state-changing logic), and it does not require creating new objects (it just mutates the existing ones). One potential drawback of the Solid approach is that `createMemo` will eagerly rerun the function every time a dependency changed, so it could result in wasted work if that memoized value is not being used anywhere currently, but I guess that could be fixed using a lazy version of `createMemo`.

> Second, I don't think we can realistically do anything that involves changing `configureStore`.

The reason I'm changing `configureStore` and `createReducer` is because I need to access the imperative logic directly (i.e. the reducers *not* wrapped by Immer), as I'm changing `dispatch` to apply this imperative logic directly to the state rather than replacing the state by a whole new state.

I'm still wondering whether there is a way to implement this idea simply as a store enhancer. Maybe it is possible to only change `createReducer` to remove Immer, then pass to `configureStore` a "reducer" that actually mutates its argument, and then have a store enhancer that replaces `dispatch`. I guess I'd need to replace `useSelector` as well.

**Reply by markerikson** [reply]
Yeah, and that's kind of my point. We can't and won't change how Redux itself works directly, which means no changes to `createStore`, `configureStore`, `createReducer`, or `createSlice`.

We _can_ try wrapping things at the React-Redux level, which is what I'm trying over in #2047 .

**Reply by guillaumebrunerie** [reply]
@markerikson Is there an easy way I can try out #2047? I tried to do `npm install reduxjs/react-redux#feature/tracking-subscriptions` (which is supposed to install a package from a github branch) but it seems to fail when building react-redux with an error I don’t understand:

```npm ERR! Internal Error: react-redux@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile```

I tried various things with yarn as well, but couldn’t make it work.

<details><summary>Details</summary>
<p>

```
$ npm install reduxjs/react-redux#feature/tracking-subscriptions
npm ERR! code 1
npm ERR! git dep preparation failed
npm ERR! command /usr/bin/node /usr/lib/node_modules/npm/bin/npm-cli.js install --force --cache=/home/fractal/.npm --prefer-offline=false --prefer-online=false --offline=false --no-progress --no-save --no-audit --include=dev --include=peer --include=optional --no-package-lock-only --no-dry-run
npm ERR! > react-redux@8.1.1 prepare
npm ERR! > yarn clean && yarn build
npm ERR! 
npm ERR! Internal Error: react-redux@workspace:.: This package doesn't seem to be present in your lockfile; run "yarn install" to update the lockfile
npm ERR!     at P0.getCandidates (/home/fractal/.npm/_cacache/tmp/git-cloneGvxekN/.yarn/releases/yarn-3.4.1.cjs:435:5145)
npm ERR!     at yf.getCandidates (/home/fractal/.npm/_cacache/tmp/git-cloneGvxekN/.yarn/releases/yarn-3.4.1.cjs:391:1264)
npm ERR!     at /home/fractal/.npm/_cacache/tmp/git-cloneGvxekN/.yarn/releases/yarn-3.4.1.cjs:439:7693
npm ERR!     at of (/home/fractal/.npm/_cacache/tmp/git-cloneGvxekN/.yarn/releases/yarn-3.4.1.cjs:390:11070)
npm ERR!     at ge (/home/fractal/.npm/_cacache/tmp/git-cloneGvxekN/.yarn/releases/yarn-3.4.1.cjs:439:7673)
npm ERR! npm WARN using --force Recommended protections disabled.
npm ERR! npm WARN ERESOLVE overriding peer dependency
npm ERR! npm WARN While resolving: @testing-library/react@12.1.5
npm ERR! npm WARN Found: react@18.0.0
npm ERR! npm WARN node_modules/react
npm ERR! npm WARN   dev react@"18.0.0" from the root project
npm ERR! npm WARN   8 more (@reduxjs/toolkit, @testing-library/jest-native, ...)
npm ERR! npm WARN 
npm ERR! npm WARN Could not resolve dependency:
npm ERR! npm WARN peer react@"<18.0.0" from @testing-library/react-12@12.1.5
npm ERR! npm WARN node_modules/@testing-library/react-12
npm ERR! npm WARN   dev @testing-library/react-12@"npm:@testing-library/react@^12" from the root project
npm ERR! npm WARN 
npm ERR! npm WARN Conflicting peer dependency: react@17.0.2
npm ERR! npm WARN node_modules/react
npm ERR! npm WARN   peer react@"<18.0.0" from @testing-library/react-12@12.1.5
npm ERR! npm WARN   node_modules/@testing-library/react-12
npm ERR! npm WARN     dev @testing-library/react-12@"npm:@testing-library/react@^12" from the root project
npm ERR! npm WARN ERESOLVE overriding peer dependency
npm ERR! npm WARN While resolving: @testing-library/react@12.1.5
npm ERR! npm WARN Found: react-dom@18.0.0
npm ERR! npm WARN node_modules/react-dom
npm ERR! npm WARN   dev react-dom@"18.0.0" from the root project
npm ERR! npm WARN   1 more (@testing-library/react)
npm ERR! npm WARN 
npm ERR! npm WARN Could not resolve dependency:
npm ERR! npm WARN peer react-dom@"<18.0.0" from @testing-library/react-12@12.1.5
npm ERR! npm WARN node_modules/@testing-library/react-12
npm ERR! npm WARN   dev @testing-library/react-12@"npm:@testing-library/react@^12" from the root project
npm ERR! npm WARN 
npm ERR! npm WARN Conflicting peer dependency: react-dom@17.0.2
npm ERR! npm WARN node_modules/react-dom
npm ERR! npm WARN   peer react-dom@"<18.0.0" from @testing-library/react-12@12.1.5
npm ERR! npm WARN   node_modules/@testing-library/react-12
npm ERR! npm WARN     dev @testing-library/react-12@"npm:@testing-library/react@^12" from the root project
npm ERR! npm WARN ERESOLVE overriding peer dependency
npm ERR! npm WARN While resolving: react-dom@17.0.2
npm ERR! npm WARN Found: react@18.0.0
npm ERR! npm WARN node_modules/react
npm ERR! npm WARN   dev react@"18.0.0" from the root project
npm ERR! npm WARN   8 more (@reduxjs/toolkit, @testing-library/jest-native, ...)
npm ERR! npm WARN 
npm ERR! npm WARN Could not resolve dependency:
npm ERR! npm WARN peer react@"17.0.2" from react-dom-17@17.0.2
npm ERR! npm WARN node_modules/react-dom-17
npm ERR! npm WARN   dev react-dom-17@"npm:react-dom@^17" from the root project
npm ERR! npm WARN 
npm ERR! npm WARN Conflicting peer dependency: react@17.0.2
npm ERR! npm WARN node_modules/react
npm ERR! npm WARN   peer react@"17.0.2" from react-dom-17@17.0.2
npm ERR! npm WARN   node_modules/react-dom-17
npm ERR! npm WARN     dev react-dom-17@"npm:react-dom@^17" from the root project
npm ERR! npm WARN ERESOLVE overriding peer dependency
npm ERR! npm WARN While resolving: react-native@0.64.4
npm ERR! npm WARN Found: react@18.0.0
npm ERR! npm WARN node_modules/react
npm ERR! npm WARN   dev react@"18.0.0" from the root project
npm ERR! npm WARN   8 more (@reduxjs/toolkit, @testing-library/jest-native, ...)
npm ERR! npm WARN 
npm ERR! npm WARN Could not resolve dependency:
npm ERR! npm WARN peer react@"17.0.1" from react-native@0.64.4
npm ERR! npm WARN node_modules/react-native
npm ERR! npm WARN   dev react-native@"^0.64.1" from the root project
npm ERR! npm WARN   3 more (@testing-library/jest-native, ...)
npm ERR! npm WARN 
npm ERR! npm WARN Conflicting peer dependency: react@17.0.1
npm ERR! npm WARN node_modules/react
npm ERR! npm WARN   peer react@"17.0.1" from react-native@0.64.4
npm ERR! npm WARN   node_modules/react-native
npm ERR! npm WARN     dev react-native@"^0.64.1" from the root project
npm ERR! npm WARN     3 more (@testing-library/jest-native, ...)
npm ERR! npm WARN ERESOLVE overriding peer dependency
npm ERR! npm WARN While resolving: react-test-renderer@17.0.2
npm ERR! npm WARN Found: react@18.0.0
npm ERR! npm WARN node_modules/react
npm ERR! npm WARN   dev react@"18.0.0" from the root project
npm ERR! npm WARN   8 more (@reduxjs/toolkit, @testing-library/jest-native, ...)
npm ERR! npm WARN 
npm ERR! npm WARN Could not resolve dependency:
npm ERR! npm WARN peer react@"17.0.2" from react-test-renderer-17@17.0.2
npm ERR! npm WARN node_modules/react-test-renderer-17
npm ERR! npm WARN   dev react-test-renderer-17@"npm:react-test-renderer@^17" from the root project
npm ERR! npm WARN 
npm ERR! npm WARN Conflicting peer dependency: react@17.0.2
npm ERR! npm WARN node_modules/react
npm ERR! npm WARN   peer react@"17.0.2" from react-test-renderer-17@17.0.2
npm ERR! npm WARN   node_modules/react-test-renderer-17
npm ERR! npm WARN     dev react-test-renderer-17@"npm:react-test-renderer@^17" from the root project
npm ERR! npm WARN deprecated source-map-url@0.4.1: See https://github.com/lydell/source-map-url#deprecated
npm ERR! npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm ERR! npm WARN deprecated rollup-plugin-terser@7.0.2: This package has been deprecated and is no longer maintained. Please use @rollup/plugin-terser
npm ERR! npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm ERR! npm WARN deprecated source-map-resolve@0.5.3: See https://github.com/lydell/source-map-resolve#deprecated
npm ERR! npm WARN deprecated w3c-hr-time@1.0.2: Use your platform's native performance.now() and performance.timeOrigin.
npm ERR! npm WARN deprecated sourcemap-codec@1.4.8: Please use @jridgewell/sourcemap-codec instead
npm ERR! npm WARN deprecated sane@4.1.0: some dependency vulnerabilities fixed, support for node < 10 dropped, and newer ECMAScript syntax/features added
npm ERR! npm WARN deprecated babel-eslint@10.1.0: babel-eslint is now @babel/eslint-parser. This package will no longer receive updates.
npm ERR! npm WARN deprecated argv@0.0.2: Package no longer supported. Contact Support at https://www.npmjs.com/support for more info.
npm ERR! npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
npm ERR! npm WARN deprecated uglify-es@3.3.9: support for ECMAScript is superseded by `uglify-js` as of v3.13.0
npm ERR! npm WARN deprecated codecov@3.8.3: https://about.codecov.io/blog/codecov-uploader-deprecation-plan/
npm ERR! npm WARN deprecated react-native@0.64.4: Issues and pull requests filed against this version are not supported. See the React Native release support policy to learn more: https://github.com/reactwg/react-native-releases#releases-support-policy
npm ERR! npm ERR! code 1
npm ERR! npm ERR! path /home/fractal/.npm/_cacache/tmp/git-cloneGvxekN
npm ERR! npm ERR! command failed
npm ERR! npm ERR! command sh -c -- yarn clean && yarn build
npm ERR! 
npm ERR! npm ERR! A complete log of this run can be found in:
npm ERR! npm ERR!     /home/fractal/.npm/_logs/2023-07-10T11_42_06_343Z-debug-0.log
```

</p>
</details>

**Reply by markerikson** [reply]
@guillaumebrunerie yes, all of our PRs are automatically built with CodeSandbox CI and can be installed using the directions listed by viewing the CSB CI job details:

https://ci.codesandbox.io/status/reduxjs/react-redux/pr/2047/builds/392030

```
# yarn 1
yarn add https://pkg.csb.dev/reduxjs/react-redux/commit/0751acc2/react-redux
# yarn 2, 3
yarn add react-redux@https://pkg.csb.dev/reduxjs/react-redux/commit/0751acc2/react-redux/_pkg.tgz
# npm
npm i https://pkg.csb.dev/reduxjs/react-redux/commit/0751acc2/react-redux
```

**Reply by guillaumebrunerie** [reply]
Thanks!

I managed to install it, but unless I am missing something it does not seem to work properly. To test it I have made a simple todo list example where we can add/remove/toggle todos, and todos have a `message: string` and a `completed: boolean`. See [here](https://codesandbox.io/s/github/guillaumebrunerie/redux-experiments) for the code.

With #2047, adding todos doesn’t update the UI and removing todos crashes the page (zombie children problem). Toggling a todo works, and it doesn’t call selectors relating to unrelated todos (which is good!), but it still unexpectedly calls the `selectTodoMessage` selector even though the message didn’t change.

You can go to the file `store/store-interface.ts` to toggle between regular React-Redux, #2047, and the Solid-inspired approach I have been exploring.

**Reply by markerikson** [reply]
@guillaumebrunerie I didn't say I knew for sure it worked right :)  that branch is something I whipped up in 4 hours on a plane ride, and it at least seemed to pass the initial unit tests. I have not had time to test it further.
